File: elixir/bob/lib/bob.ex
defmodule Bob do
  @spec hey(String.t()) :: String.t()

  def hey(input) do
    trimmed = String.trim(to_string(input))

    cond do
      String.upcase(trimmed) == trimmed and String.downcase(trimmed) != trimmed and
          String.last(trimmed) == "?" ->
        "Calm down, I know what I'm doing!"

      String.upcase(trimmed) == trimmed and String.downcase(trimmed) != trimmed ->
        "Whoa, chill out!"

      String.last(trimmed) == "?" ->
        "Sure."

      String.trim(trimmed) == "" ->
        "Fine. Be that way!"

      true ->
        "Whatever."
    end
  end

  # @spec yell_question?(String.t()) :: boolean
  # defp yell_question?(s) do
  #   s |> yell?() and s |> question?()
  # end
  #
  # @spec yell?(String.t()) :: boolean
  # defp yell?(s) do
  #   trimmed = String.trim(s)
  #
  #   trimmed == String.upcase(trimmed) and String.match?(trimmed, ~r/[A-Z]/)
  # end
  #
  # @spec question?(String.t()) :: boolean
  # defp question?(s) do
  #   s |> String.trim() |> String.ends_with?("?")
  # end
  #
  # @spec silence?(String.t()) :: boolean
  # defp silence?(s) do
  #   s |> String.trim() == ""
  # end
end

File: scala/bob/src/main/scala/Bob.scala
object Bob:
  def yell(msg: String): Boolean =
    msg.toUpperCase == msg && msg.toLowerCase != msg

  def question(msg: String): Boolean =
    msg.trim.endsWith("?")

  def yellAndQuestion(msg: String): Boolean =
    yell(msg) && question(msg)

  def silence(msg: String): Boolean =
    msg.trim == ""

  def response(msg: String): String =
    if yellAndQuestion(msg) then "Calm down, I know what I'm doing!"
    else if yell(msg) then "Whoa, chill out!"
    else if question(msg) then "Sure."
    else if silence(msg) then "Fine. Be that way!"
    else "Whatever."
